File: companion/ZippieCompanionKit/Sources/ZippieCompanionKit/BondStatus.swift
import Foundation

/// A snapshot of the zippie bond, decoded from the router's /api/status.
///
/// Read-only and defensive: the console is a developer surface that changes
/// shape as the agent evolves, and an app that crashes because a field was
/// renamed is worse than one that shows a dash. Every field is optional.
public struct BondStatus: Decodable, Sendable, Equatable {
    public struct Path: Decodable, Sendable, Equatable {
        public let name: String?
        public let state: String?
        public let effectiveWeight: Int?
        public let rttMs: Double?
        public let lossPct: Double?
        /// What a human called this link. The `name` is an internal id and a
        /// poor thing to read on a dashboard at speed.
        public let label: String?
        /// Absent when the leg is configured but not physically present - a
        /// dongle that is unplugged. That is how a leg gets filtered out of
        /// the bond rather than padding it as a permanent "down" row.
        public let interface: String?
        public let txBytes: Int?
        public let rxBytes: Int?
        /// The TRANSPORT's own per-link counters.
        ///
        /// THESE ARE THE REAL ONES IN PACKET MODE. tx_bytes/rx_bytes come from
        /// /sys/class/net/<wg_iface>, and packet mode has no per-leg wg
        /// interface - so they are hard zero for every leg no matter how much
        /// it has carried. A leg that had moved 30 MB drew an EMPTY traffic
        /// bar, which is precisely the "healthy but carrying nothing" lie this
        /// app exists to expose, told in reverse.
        public let linkTxBytes: Int?
        public let linkRxBytes: Int?
        /// The router's own explanation, e.g. "healthy, held out of bond until
        /// proven (2.5/8)". Worth surfacing verbatim: it is the difference
        /// between a diagnosis and a red dot.
        public let lastError: String?
        /// For companion legs only: the address:port the router dials to reach
        /// that phone. A phone matches this against its own wifi address and
        /// listen port to know which leg is itself - see infra#2247. Empty on
        /// physical legs, and absent entirely on a router too old to publish
        /// it, which is why it is optional rather than defaulted here.
        public let relayEndpoint: String?
        /// The hard failover gate. A tier-2 leg carries NOTHING while any
        /// tier-1 leg is alive - it is held in reserve, not broken, and those
        /// two look identical from outside unless the UI is told which.
        public let tier: Int?
        /// A deliberate ceiling in kbit/s; 0 or absent means uncapped. Without
        /// this a throttled leg reads as a slow one.
        public let maxKbps: Int?
        public let costClass: String?
        public let monthlyCapGB: Double?
        public let usageGB: Double?
        public let overSoftLimit: Bool?
        /// Whether the transport actually holds a link for this leg.
        ///
        /// NOT THE SAME AS HAVING WEIGHT, and that difference is why this
        /// exists. A tier-gated leg keeps whatever weight the policy last
        /// computed - the number is real, it is simply not being used - so
        /// deciding "carrying" from weight showed four legs carrying while the
        /// transport held exactly one.
        public let inBond: Bool?
        /// The leg has transmitted and has NEVER been answered (#204).
        /// Categorically different from `degraded`, which is where a leg lands
        /// when it used to work and got worse.
        public let neverHandshaked: Bool?
        /// Usable uplinks this leg's pattern matched and no leg took (#212) -
        /// a link that is working and invisible.
        public let shadowedInterfaces: [String]?

        enum CodingKeys: String, CodingKey {
            case name, state, label, interface
            case effectiveWeight = "effective_weight"
            case rttMs = "rtt_ms"
            case lossPct = "loss_pct"
            case txBytes = "tx_bytes"
            case rxBytes = "rx_bytes"
            case linkTxBytes = "link_tx_bytes"
            case linkRxBytes = "link_rx_bytes"
            case lastError = "last_error"
            // Optionals, so a router predating these fields decodes exactly as
            // it does today.
            case neverHandshaked = "never_handshaked"
            case shadowedInterfaces = "shadowed_interfaces"
            case relayEndpoint = "relay_endpoint"
            case tier
            case maxKbps = "max_kbps"
            case costClass = "cost_class"
            case monthlyCapGB = "monthly_cap_gb"
            case usageGB = "usage_gb"
            case overSoftLimit = "over_soft_limit"
            case inBond = "in_bond"
        }

        /// Held in reserve by the tier gate rather than failed.
        ///
        /// THE DISTINCTION THE WHOLE STATUS SCREEN TURNS ON. A reserve leg
        /// reports no traffic because it is DOING ITS JOB - a cheap SIM kept
        /// for the day everything else is down. Drawing it the same as a
        /// broken leg trains the reader to ignore the one signal that matters.
        ///
        /// Decided against the ACTIVE tier, not against tier 1: if the bond
        /// has fallen to tier 2, a tier-2 leg is live and only tier 3 is still
        /// in reserve.
        public func isHeldInReserve(activeTier: Int?) -> Bool {
            guard let tier, let activeTier else { return false }
            return tier > activeTier
        }

        /// Bytes this leg actually carried, from whichever counter is real.
        ///
        /// Link counters first because they are the packet-mode truth; the
        /// sysfs pair is the route-mode fallback. Falling back rather than
        /// preferring one keeps both datapaths honest without the app needing
        /// to know which is running.
        public var carriedTx: UInt64 { UInt64(max(0, linkTxBytes ?? txBytes ?? 0)) }
        public var carriedRx: UInt64 { UInt64(max(0, linkRxBytes ?? rxBytes ?? 0)) }

        /// True when this leg is a phone rather than a physical uplink.
        public var isCompanion: Bool { !((relayEndpoint ?? "").isEmpty) }

        /// Present in the bond at all. A path with neither an interface nor a
        /// relay endpoint is configuration, not a connection.
        public var isPresent: Bool {
            !((interface ?? "").isEmpty) || isCompanion
        }

        /// A leg is only carrying if it has weight. `state == "up"` is not
        /// enough - the anti-flap gate holds recovered legs at weight 0 while
        /// they prove themselves, and showing those as carrying would
        /// contradict the router's own console.
        /// Carrying means the transport holds a link for it AND it has weight.
        ///
        /// Weight alone is not enough - see inBond. A router too old to publish
        /// membership falls back to weight, which is the previous behaviour and
        /// wrong only in the tier-gated case that older agent could not produce.
        public var isCarrying: Bool {
            guard (effectiveWeight ?? 0) > 0 else { return false }
            return inBond ?? true
        }

        /// A companion leg that has never had a byte back, and never measured
        /// a round trip, is not a degraded connection - it is an address with
        /// nothing at it. The router keeps a leg configured for a phone that
        /// has left the network and keeps sending to it, so this state is
        /// normal and must not be dressed up as a link having a bad day.
        public var neverAnswered: Bool {
            isCompanion && (linkRxBytes ?? rxBytes ?? 0) == 0 && rttMs == nil
        }

        /// The state in ONE WORD, because "is it degraded, or is it one of two"
        /// was not answerable from this screen.
        public var stateWord: String {
            if isCarrying { return state == "degraded" ? "carrying, degraded" : "carrying" }
            if neverAnswered { return "not connected" }
            if inBond == false { return "not in the bond" }
            switch state {
            case "up":       return "up, not carrying"
            case "degraded": return "degraded"
            case "down":     return "down"
            default:         return "idle"
            }
        }
    }

    public let mode: String?
    public let datapath: String?
    public let primary: String?
    public let paths: [Path]?

    /// The lowest tier with a carrying leg - the tier the bond is currently
    /// running on. Nil when nothing carries at all, which is a different and
    /// much worse state than "running on a lower tier".
    public var activeTier: Int? {
        (paths ?? []).filter(\.isCarrying).compactMap(\.tier).min()
    }

    public var carryingCount: Int { (paths ?? []).filter(\.isCarrying).count }
    public var totalCount: Int { (paths ?? []).count }
}

public struct BondStatusError: Error, Equatable, Sendable {
    public let message: String
    public init(_ m: String) { message = m }
}

public enum BondStatusClient {
    /// Fetch and decode the router's status. Deliberately takes the full URL so
    /// the app can point at a tailnet name, a LAN IP, or the public console
    /// without the Kit knowing anything about the topology.
    /// The session is injectable so the APP can pass a Datadog-instrumented
    /// one. dd-sdk-ios only traces a task whose delegate matches the class it
    /// was told to instrument, and URLSession.shared has no delegate at all -
    /// so tracing configured against the shared session is silently inert.
    /// Defaulting to .shared keeps the Kit usable on a bare toolchain.
    public static func fetch(url: URL,
                             timeout: TimeInterval = 8,
                             session: URLSession = .shared) async -> Result<BondStatus, BondStatusError> {
        var req = URLRequest(url: url)
        req.timeoutInterval = timeout
        req.cachePolicy = .reloadIgnoringLocalCacheData
        do {
            let (data, resp) = try await session.data(for: req)
            if let http = resp as? HTTPURLResponse, http.statusCode != 200 {
                return .failure(BondStatusError("HTTP \(http.statusCode)"))
            }
            return .success(try JSONDecoder().decode(BondStatus.self, from: data))
        } catch {
            return .failure(BondStatusError(error.localizedDescription))
        }
    }
}

File: companion-android/app/src/main/java/app/zippie/companion/BondStatus.kt
package app.zippie.companion

import org.json.JSONException
import org.json.JSONObject

/**
 * A snapshot of the zippie bond, decoded from the router's /api/status.
 *
 * Read-only and defensive, for the same reason BondStatus.swift is: the console
 * is a developer surface that changes shape as the agent evolves, and an app
 * that crashes because a field was renamed is worse than one that says it does
 * not know. EVERY FIELD IS NULLABLE, and null means "the router did not tell
 * us", which the UI must render as unknown rather than as zero.
 *
 * org.json rather than a reflection-based decoder on purpose. Its optInt/optLong
 * accessors return 0 for a missing key, which is precisely the failure this
 * type exists to prevent, so nothing here calls them - see [intOrNull] and
 * friends below.
 */
data class BondStatus(
    val mode: String?,
    val datapath: String?,
    val primary: String?,
    /**
     * Null means the payload had no `paths` key at all - a shape we do not
     * understand. An empty list means the router genuinely lists no links.
     * Those are different facts and the screen says different things about them.
     */
    val paths: List<Path>?,
) {
    data class Path(
        val name: String?,
        /** What a human called this link. `name` is an internal id and a poor
         *  thing to read on a dashboard at speed. */
        val label: String?,
        val state: String?,
        /**
         * The hard failover gate. A leg on a higher tier carries NOTHING while
         * a lower tier is alive - it is held in reserve, not broken, and those
         * two look identical from outside unless the UI is told which.
         */
        val tier: Int?,
        val effectiveWeight: Int?,
        val rttMs: Double?,
        val lossPct: Double?,
        /**
         * The legacy per-leg counters. In `datapath: "packet"` these stay at 0
         * on every leg while traffic flows - see [carriedTxBytes].
         */
        val txBytes: Long?,
        val rxBytes: Long?,
        /**
         * The packet datapath's own per-link counters. Verified against the live
         * router 2026-08-05: hotspot reported tx_bytes 0 / rx_bytes 0 while
         * link_tx_bytes 11018643 and link_rx_bytes 31703754, and the transport's
         * reassembly counter agreed with the latter pair.
         */
        val linkTxBytes: Long?,
        val linkRxBytes: Long?,
        /** Absent when the leg is configured but not physically present - an
         *  unplugged dongle. */
        val interfaceName: String?,
        /** The router's own explanation, e.g. "awaiting transport". Worth
         *  surfacing verbatim: it is the difference between a diagnosis and a
         *  red dot. */
        val lastError: String?,
        /** The leg has transmitted and has NEVER been answered (#204). Not the
         *  same as degraded: that is a leg which used to work and got worse.
         *  This one is pointed at something that is not listening. */
        val neverHandshaked: Boolean?,
        /** Usable uplinks this leg's pattern matched and no leg took (#212) -
         *  a link that is working and invisible. */
        val shadowedInterfaces: List<String>?,
        /**
         * For companion legs only: the address:port the router dials to reach
         * that phone. A phone matches this against its own wifi address and
         * listen port to know which leg is itself - see LegIdentity.
         */
        val relayEndpoint: String?,
        /** A deliberate ceiling in kbit/s; 0 or absent means uncapped. Without
         *  this a throttled leg reads as a slow one. */
        val maxKbps: Int?,
        val costClass: String?,
        val monthlyCapGb: Double?,
        val usageGb: Double?,
        val overSoftLimit: Boolean?,
        /**
         * Whether the transport actually holds a link for this leg.
         *
         * NOT THE SAME AS HAVING WEIGHT, and that difference is why this field
         * exists. A tier-gated leg keeps whatever weight the policy last
         * computed - the number is real, it is simply not being used - so
         * deciding "carrying" from weight showed four legs carrying while the
         * transport held exactly one. That was a real bug on iOS.
         */
        val inBond: Boolean?,
    ) {
        /**
         * Bytes this leg has actually moved, or null when nothing measured it.
         *
         * Prefers the packet datapath's link counters because the older
         * tx_bytes/rx_bytes pair is hard 0 in that mode: reading those would
         * draw a leg that has carried 30 MB as having carried nothing, which is
         * the exact class of lie this app exists to avoid.
         */
        val carriedTxBytes: Long? get() = linkTxBytes ?: txBytes
        val carriedRxBytes: Long? get() = linkRxBytes ?: rxBytes

        /**
         * Carrying means the transport holds a link for it AND it has weight.
         *
         * Weight alone is not enough - see [inBond]. A router too old to publish
         * membership falls back to weight, which is the previous behaviour and
         * wrong only in the tier-gated case that older agent could not produce.
         */
        val isCarrying: Boolean
            get() = (effectiveWeight ?: 0) > 0 && (inBond ?: true)

        /**
         * Held back by the tier gate rather than failed.
         *
         * THE DISTINCTION THE WHOLE STATUS SCREEN TURNS ON. A reserve leg
         * reports no traffic because it is DOING ITS JOB - a cheap SIM kept for
         * the day everything else is down. Drawing it the same as a broken leg
         * trains the reader to ignore the one signal that matters.
         *
         * Decided against the ACTIVE tier, not against tier 1: if the bond has
         * already fallen to tier 2, a tier-2 leg is live and only tier 3 is
         * still in reserve.
         */
        fun isHeldInReserve(activeTier: Int?): Boolean {
            if (tier == null || activeTier == null) return false
            return tier > activeTier
        }

        /** True when this leg is a phone rather than a physical uplink. */
        val isCompanion: Boolean get() = !relayEndpoint.isNullOrEmpty()

        /** Present in the bond at all. A path with neither an interface nor a
         *  relay endpoint is configuration, not a connection. */
        val isPresent: Boolean get() = !interfaceName.isNullOrEmpty() || isCompanion

        /**
         * The state in ONE WORD, because "is it degraded, or is it one of two"
         * was not answerable from the screen otherwise.
         */
        val stateWord: String
            get() {
                if (isCarrying) return if (state == "degraded") "carrying, degraded" else "carrying"
                if (inBond == false) return "not in the bond"
                return when (state) {
                    "up" -> "up, not carrying"
                    "degraded" -> "degraded"
                    "down" -> "down"
                    else -> "idle"
                }
            }
    }

    /**
     * The lowest tier with a carrying leg - the tier the bond is running on.
     * Null when nothing carries at all, which is a different and much worse
     * state than "running on a lower tier".
     */
    val activeTier: Int?
        get() = (paths ?: emptyList()).filter { it.isCarrying }.mapNotNull { it.tier }.minOrNull()

    val carryingCount: Int get() = (paths ?: emptyList()).count { it.isCarrying }
    val totalCount: Int get() = (paths ?: emptyList()).size

    companion object {
        /**
         * Throws [JSONException] on a payload that is not JSON at all. A body
         * that parses but carries none of the fields we expect decodes to a
         * BondStatus full of nulls, which the UI renders as "the router
         * answered with something we do not understand" - still better than a
         * crash, and it cannot be mistaken for a healthy bond.
         */
        @Throws(JSONException::class)
        fun decode(payload: String): BondStatus {
            val root = JSONObject(payload)
            val pathsArray = if (root.isNull("paths")) null else root.optJSONArray("paths")
            val paths = pathsArray?.let { array ->
                (0 until array.length()).mapNotNull { i ->
                    array.optJSONObject(i)?.let(::decodePath)
                }
            }
            return BondStatus(
                mode = root.stringOrNull("mode"),
                datapath = root.stringOrNull("datapath"),
                primary = root.stringOrNull("primary"),
                paths = paths,
            )
        }

        private fun decodePath(o: JSONObject) = Path(
            name = o.stringOrNull("name"),
            label = o.stringOrNull("label"),
            state = o.stringOrNull("state"),
            tier = o.intOrNull("tier"),
            effectiveWeight = o.intOrNull("effective_weight"),
            rttMs = o.doubleOrNull("rtt_ms"),
            lossPct = o.doubleOrNull("loss_pct"),
            txBytes = o.longOrNull("tx_bytes"),
            rxBytes = o.longOrNull("rx_bytes"),
            linkTxBytes = o.longOrNull("link_tx_bytes"),
            linkRxBytes = o.longOrNull("link_rx_bytes"),
            interfaceName = o.stringOrNull("interface"),
            lastError = o.stringOrNull("last_error"),
            // ABSENT-SAFE. A router that predates these fields simply does not
            // send them, and must render exactly as it does today.
            neverHandshaked = if (o.has("never_handshaked")) o.optBoolean("never_handshaked") else null,
            shadowedInterfaces = o.optJSONArray("shadowed_interfaces")?.let { arr ->
                (0 until arr.length()).mapNotNull { arr.optString(it).takeIf { s -> s.isNotEmpty() } }
            },
            relayEndpoint = o.stringOrNull("relay_endpoint"),
            maxKbps = o.intOrNull("max_kbps"),
            costClass = o.stringOrNull("cost_class"),
            monthlyCapGb = o.doubleOrNull("monthly_cap_gb"),
            usageGb = o.doubleOrNull("usage_gb"),
            overSoftLimit = o.booleanOrNull("over_soft_limit"),
            inBond = o.booleanOrNull("in_bond"),
        )
    }
}

// The accessors below all answer "absent" with null instead of with a zero or a
// false. org.json's opt* family answers with 0 / false / "" and that is how an
// unmeasured RTT becomes a confident "0 ms" on screen.

internal fun JSONObject.stringOrNull(key: String): String? {
    if (!has(key) || isNull(key)) return null
    val s = optString(key, "")
    return s.ifEmpty { null }
}

internal fun JSONObject.intOrNull(key: String): Int? {
    if (!has(key) || isNull(key)) return null
    return try {
        getInt(key)
    } catch (e: JSONException) {
        null
    }
}

internal fun JSONObject.longOrNull(key: String): Long? {
    if (!has(key) || isNull(key)) return null
    return try {
        getLong(key)
    } catch (e: JSONException) {
        null
    }
}

internal fun JSONObject.doubleOrNull(key: String): Double? {
    if (!has(key) || isNull(key)) return null
    return try {
        getDouble(key)
    } catch (e: JSONException) {
        null
    }
}

internal fun JSONObject.booleanOrNull(key: String): Boolean? {
    if (!has(key) || isNull(key)) return null
    return try {
        getBoolean(key)
    } catch (e: JSONException) {
        null
    }
}
